include/bluetooth/enocean.h

```diff
@@ -0,0 +1,201 @@
+	 *
+	 *  @param device Device that got commissioned.
+	 */
+	void (*commissioned)(struct bt_enocean_device *device);
+
+	/** @brief Callback for EnOcean devices being loaded from
+	 *         persistent storage.
+	 *
+	 *  This callback is called for every commissioned EnOcean device
+	 *  recovered from persistent storage.
+	 *
+	 *  @param device Device that was loaded.
+	 */
+	void (*loaded)(struct bt_enocean_device *device);
+};
+
+/** @brief Visitor callback for @ref bt_enocean_foreach.
+ *
+ *  @param dev       EnOcean device
+ *  @param user_data User data passed to @ref bt_enocean_foreach.
+ */
+typedef void (*bt_enocean_visitor_t)(struct bt_enocean_device *dev,
+				     void *user_data);
+
+/** @brief Initialize the EnOcean subsystem.
+ *
+ *  @param cb Set of callbacks to receive events with.
+ */
+void bt_enocean_init(const struct bt_enocean_callbacks *cb);
+
+/** @brief Enable automatic commissioning.
+ *
+ *  This function must be called for the module to start accepting
+ *  commissioning packets.
+ */
+void bt_enocean_commissioning_enable(void);
+
+/** @brief Disable automatic commissioning.
+ *
+ *  Automatic commissioning will be disabled until
+ *  @ref bt_enocean_commissioning_enable is called. Manual commissioning
+ *  through @ref bt_enocean_commission is unaffected.
+ */
+void bt_enocean_commissioning_disable(void);
+
+/** @brief Manually commission an EnOcean device.
+ *
+ *  Allows devices to be commissioned based on out of band data, like NFC or QR
+ *  codes.
+ *
+ *  @param addr Address of the commissioned device.
+ *  @param key  Encryption key for the commissioned device.
+ *  @param seq  Most recent sequence number of the device.
+ *
+ *  @retval 0          The device has been successfully commissioned.
+ *  @retval -EEXIST    This device has already been commissioned.
+ *  @retval -ENOMEM    There's not enough space to commission a new device.
+ *  @retval -ECANCELED The commissioning callback rejected the device.
+ *  @retval -EBUSY     Failed allocating persistent storage data.
+ */
+int bt_enocean_commission(const u8_t addr[6], const u8_t key[16], u32_t seq);
+
+/** @brief Forget an EnOcean device.
+ *
+ *  Removes the EnOcean device's security information, freeing up space for new
+ *  devices. May be used directly in the
+ *  @ref bt_enocean_callbacks::commissioning callback to reject a device before
+ *  it gets commissioned.
+ *
+ *  @param addr Address of the device to forget.
+ *
+ *  @retval 0       Successfully decommissioned the EnOcean device.
+ *  @retval -ENOENT The device doesn't exist.
+ */
+int bt_enocean_decommission(const u8_t addr[6]);
+
+/** @brief Call the given callback for each commissioned enocean device.
+ *
+ *  @param cb        Callback to call for each device
+ *  @param user_data User data to pass to the callback.
+ *
+ *  @return The number of devices visited.
+ */
+u32_t bt_enocean_foreach(bt_enocean_visitor_t cb, void *user_data);
```

Comment:
Is this the visitor pattern? The foreach functions usually don't typedef the callback. I'm wondering if the would be best to do the same here.

Reply:
:( It's objectively better though? Less code duplication and documentation for the callback parameters

Reply:
Sure, just wondering about the typedef name. maybe `bt_onocean_foreach_cb_t`?